### PR TITLE
catalog: Remove stash dependency from init

### PR DESCRIPTION
**Pull request by jkosh44**
This commit removes the catalog initialization dependency on the stash. It does that by updating the catalog initialization to use a catalog transaction instead of a stash transaction.

Works towards resolving #20953

### Motivation
This PR adds a known-desirable feature.

### Tips for reviewer
Hide whitespace.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
@mjibson I switched you as the reviewer for @ParkMyCar since he wrote the initialization function. Feel free to take a look as well if you want.

File: src/catalog/src/transaction.rs [hunk 1]
```diff
@@ -281,19 +281,28 @@ impl<'a> Transaction<'a> {
         privileges: Vec<MzAclItem>,
     ) -> Result<DatabaseId, Error> {
         let id = self.get_and_increment_id(DATABASE_ID_ALLOC_KEY.to_string())?;
+        // TODO(parkertimmerman): Support creating databases in the System namespace.
+        let id = DatabaseId::User(id);
+        self.insert_database(id, database_name, owner_id, privileges)?;
+        Ok(id)
+    }
+
+    pub(crate) fn insert_database(
```
Comment (ParkMyCar) [inline src/catalog/src/transaction.rs:290]: should we get rid of `fn insert_user_database(...)`? As well as the equiv functions for schema and role?

**Review by ParkMyCar — APPROVED** [review]
Nice! This looks great, the initialization flow is much cleaner now.

You don't need to add it here, but it might be nice if we added a snapshot test with the [insta](https://docs.rs/insta/1.33.0/insta/) crate to assert the initial state of the Catalog, what do you think?

**Comment by jkosh44** [comment]
> You don't need to add it here, but it might be nice if we added a snapshot test with the [insta](https://docs.rs/insta/1.33.0/insta/) crate to assert the initial state of the Catalog, what do you think?

Good idea: https://github.com/MaterializeInc/materialize/issues/22278

File: src/catalog/src/transaction.rs (hunk 1 above)
Comment (jkosh44) [reply src/catalog/src/transaction.rs:290]: Those are still useful for catalog transactions which (almost) only ever create user objects.

**Pull request merged**